src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -637,6 +637,12 @@ private static object SplitWithPredicate(ExecutionContext context, IScriptExtent
             {
                 List<string> split = new List<String>();
```

Comment:
It seems we could move this to line 653 as
```c#
List<string> split = new List<String>(limit); 
```

Reply:
I'd keep it without specifying capacity up front, because you'd still need to support unbounded capacity in the special case of (`limit == 0`) and the (very common) case of _no `Max-substrings` specified_.

Reply:
@iSazonov What if someone put `1000000` as `limit`? I do not think you should pre-allocate so much memory.

Reply:
I skipped ` limit == 0 `
Closed.

Reply:
Why shouldn't we specify the limit in the ctor? 
I'd gladly have powershell allocate the list once to the size I had given than have the list being reallocated and copied many times when the value is larger?
If someone specifies `1000000` they will get a list of roughly 8MB, which would soon be garbage collected, but with a single allocation, instead of the 19 we would get with the 0 default capacity.

Reply:
Fair point. We could also allocate the list with capacity `Min(limit, item.Length +2)` (unless there's some negative-width split pattern I can't think of)

Reply:
I agree - it is good. Reallocation is much worse.

Reply:
Forget about that `Min(limit, item.Length +2)` comment above, I just realized we're talking about regex with capture support, including `'abcde' -split '((?=.)(.)(?<=.))'` similarly terrifying things.

I've added the suggested changes + comments for the `List.Reverse()` call in https://github.com/PowerShell/PowerShell/pull/5125/commits/c423d8d6c998dd01ca9e7e2c54752d635bd7dd78

Reply:
I still like your thought to use `Min(limit, item.Length +2)`. It covers the common case sufficiently and protects against complete stupidity.

Reply:
Just had a look at the [`List<T>` source](https://referencesource.microsoft.com/#mscorlib/system/collections/generic/list.cs,51), the default capacity reallocation behavior (multiply by 2) actually fits our needs perfectly if we do `Min(limit, item.Length + 2)` - matches worst case with no captures, and each nested capture-group added will at most require 1 extra re-allocation.

Reply:
[CoreFX link on EnsureCapacity](https://github.com/dotnet/corefx/blob/bffef76f6af208e2042a2f27bc081ee908bb390b/src/System.Collections/src/System/Collections/Generic/List.cs#L433)